1-Draft/RFCNNNN-Loading-Module-Into-Isolated-AssemblyLoadContext.md

```diff
@@ -0,0 +1,183 @@
+---
+RFC: RFC0000
+Author: Dongbo Wang
+Status: Draft
+SupercededBy: N/A
+Version: 1.0
+Area: Microsoft.PowerShell.Core
+Comments Due: 4/10/2019
+Plan to implement: No
+---
+
+# Loading Modules into Isolated AssemblyLoadContext
+
+Assembly isolation is a missing feature in PowerShell.
+Today, a PowerShell process loads all assemblies in the default load context.
+When an assembly is already loaded, it's not allowed to load the same assembly again with a different version.
+This results in the PowerShell issue [#2083](https://github.com/PowerShell/PowerShell/issues/2083).
+Basically, when two modules depend on different versions of the same assembly,
+if the module that depends on the lower version is loaded first,
+the other module cannot be loaded in the same session anymore.
+
+There are two layers of assembly isolation that we can consider to support
+
+- Runspace level: Load PowerShell assemblies into a custom `AssemblyLoadContext`,
+  then create a Runspace via Reflection within that `AssemblyLoadContext`.
+  So that Runspace works with its own set of PowerShell assemblies in that load context,
+  and all assemblies loaded in that Runspace also get loaded into that load context
+  (Except for `Assembly.LoadFrom` and `Assembly.LoadFile`.
+   `LoadFrom` always loads an assembly to the default load context;
+   `LoadFile` always loads an assembly into a new load context).
+
+- Module level: Load a module into a custom `AssemblyLoadContext`,
+  the module required assemblies as well as all the assemblies loaded by `Add-Type` in the module script will be loaded into that load context.
+
+Note: _This RFC focuses on the module-level isolation, **but it's written with the purpose to be withdrawn**._
+      _While researching the feasibility of this idea, I realized that it's impossible to have it well supported due to the challenges discussed below._
+      _This RFC serves as a summary of the research to explain the reasons why we will not proceed with this idea._
+
+## Motivation
+
+    As a PowerShell user, I'm able to use modules that depend on different versions of the same assembly in the same session.
+
+## Specification
+
+### Import-Module -Isolated
+
+The idea is simple: Add the switch parameter `-Isolated` to `Import-Module`.
```

Comment:
Regardless of how we implement this, I don't think we can require the user to use a switch to specify isolation.  The majority of users won't understand why this is needed nor when to use it.  Only in developer scenarios would such a switch make sense.  For regular users, it needs to be transparent.

Reply:
>  For regular users, it needs to be transparent.

That makes sense, and I think `AssemblyLoadContext.AssemblyResolve` event might be a better fit for solving the conflicting `NewtonSoft.Json` issue, in a transparent way.
As for the module isolation, it won't work. The introduced type identity issue will make it impossible to be transparent to the user.